### PR TITLE
Issue #3260416 by vnech: As a user I should be able to follow another users

### PR DESCRIPTION
**Pull request by nechai**
## Problem
As a LU User I should be able to follow another users. 

That means I should be able:

1. Follow any user
2. Unfollow any user

## Solution
Add a new custom module allows follow/unfollow other users

## Issue tracker
- https://getopensocial.atlassian.net/browse/YANG-6955
- https://www.drupal.org/project/social/issues/3260416
- https://www.drupal.org/project/socialblue/issues/3268959

## How to test
- [ ] Install the module `social_follow_user` 
- [ ] Login as LU
- [ ] Visit the profile page any user (you should see "Follow" button and be able "follow/unfollow" this user)

## Release notes
Make possible to"follow/unfollow" any user.

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by rollki** [comment]
LGTM.

**Review by rollki — APPROVED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

File: modules/social_features/social_follow_user/social_follow_user.install [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Lifecycle functions for the "Social Follow User" module.
+ */
+
+/**
+ * Implements hook_install().
+ *
+ * Perform actions related to the installation of social_follow_user.
+ */
+function social_follow_user_install(): void {
+  // Grant the default permissions for this feature.
```
Comment (rollki) [inline modules/social_features/social_follow_user/social_follow_user.install:14]: It's can be more compact, e.g.:
```
$roles = [
'verified',
'contentmanager',
'sitemanager',
];

And foreach where you call `user_role_grant_permissions`
```

**Comment by rollki** [comment]
@nechai There are one thing in code that we can change, please check it out.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by rollki** [comment]
LGTM, @OH17L Please  go ahead with your review here :)

**Review by OH17L — CHANGES_REQUESTED** [review]

File: modules/social_features/social_follow_user/social_follow_user.info.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+name: 'Social Follow User'
+description: 'Provides "Follow User" flag type and related functionality.'
+type: module
+core_version_requirement: ^8.8 || ^9
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.info.yml]: fixed

File: modules/social_features/social_follow_user/social_follow_user.info.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+dependencies:
+  - drupal:user
+  - flag:flag
+package: Social
+hidden: true
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.info.yml]: fixed

File: modules/social_features/social_follow_user/social_follow_user.install [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/**
+ * @file
+ * Lifecycle functions for the "Social Follow User" module.
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.install]: fixed

File: modules/social_features/social_follow_user/social_follow_user.module [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Implements hook_follow_user_flag_action_access().
+ */
+function social_follow_user_flag_action_access(string $action, FlagInterface $flag, AccountInterface $account, EntityInterface $flaggable = NULL): AccessResultInterface {
+  if ($flaggable instanceof ProfileInterface) {
+    // User should not be able to follow own profile.
+    return AccessResult::forbiddenIf($flaggable->getOwnerId() === $account->id());
+  }
+
+  return AccessResult::neutral();
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.module]: fixed

File: modules/social_features/social_follow_user/social_follow_user.install [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  // Grant the default permissions for this feature.
+  $role_ids = ['verified', 'contentmanager', 'sitemanager'];
+  foreach ($role_ids as $role_id) {
+    user_role_grant_permissions(
+      $role_id,
+      [
+        'flag follow_user',
+        'unflag follow_user',
+      ]
+    );
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.install]: fixed

File: modules/social_features/social_follow_user/social_follow_user.module [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * Implements hook_entity_view_display_alter().
+ */
+function social_follow_user_entity_view_display_alter(EntityViewDisplayInterface $display, array $context): void {
+  // Follow button appears on listed profile display modes.
+  $allowed_display_modes = ['statistic'];
+
+  // Allow other modules to change the list.
+  \Drupal::moduleHandler()->alter('social_follow_user_profile_modes', $allowed_display_modes);
+
+  if ($context['entity_type'] === 'profile' && !in_array($context['view_mode'], $allowed_display_modes)) {
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_follow_user/social_follow_user.module]: fixed

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai 
All look good, but we need to fix few things before we can merge this:

1. Please make this module a hidden for now
2. Add more information to issue on d.org
3. Rebase commits

Thank you!

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2712-wkqlsore5fnhjiy4wururdrefvlxev1j.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f0372d046a4b8d29146ab9

**Review by ribel — APPROVED** [review]
Now all look good, so I am going to merge this PR.
Thank you @nechai!

**Pull request merged**